File: code-awesome/django-react/super_dating/backend/src/pages/message.py
from django.shortcuts import render, HttpResponse, render_to_response
from django.template import Context, Template
from django.contrib import messages
from django.contrib import staticfiles
import os
import hashlib
import random
from datetime import datetime
import json

def message_view(request):
    """
    Handles messages - sending, receiving, and viewing.

    This view demonstrates several features:
    - Sending messages (basic)
    - Receiving messages (displaying)
    - Basic chat functionality (simulated)
    - Public forum (placeholder)

    It also includes a demonstration of content injection vulnerabilities,
    though this is a simplified example for educational purposes.

    """
    # -------------------- MESSAGE SENDING --------------------
    if request.method == 'POST':
        # Basic message sending - in a real app, you'd sanitize and validate
        # all input, and likely store this data in a database
        message_text = request.POST.get('message')
        if message_text:
            messages.info(request, "Message Sent!")  # Use Django's messaging
            # In a real application, you'd store this message in a database
            # and associate it with the sender and recipient (if applicable)
        else:
            messages.error(request, "Message text is empty!") #error message if there is no text

        return render_to_response('message_view.html', context={}, request=request) #simple render, no need for the html
        #return HttpResponse("Message sent successfully!")


    # -------------------- MESSAGE RECEIVING --------------------
    # Simulate receiving messages (this is a placeholder)
    # In a real app, you'd fetch messages from a database
    # based on sender, recipient, etc.
    #Example
    #messages = [
    #    {"sender": "User1", "message": "Hello there!", "timestamp": datetime.now()},
    #    {"sender": "User2", "message": "How are you?", "timestamp": datetime.now()}
    #]

    # Example using context to pass data for the template
    #context = {'messages': messages} # this is how you will do that in real life

    #return render_to_response('message_view.html', context=context, request=request)

    # -------------------- PUBLIC FORUM (PLACEHOLDER) --------------------
    # Placeholder for a public forum - usually involves a more complex
    # database structure and user authentication.
    # You could display a list of recent forum posts here.
    #context = {'forum_posts': [
    #    {"author": "Admin", "post": "Welcome to the forum!"},
    #    {"author": "User3", "post": "Interesting discussion..."}
    #]}
    #return render_to_response('message_view.html', context=context, request=request)



    # -------------------- DEFAULT RESPONSE --------------------
    return render_to_response('message_view.html', context={}, request=request) # render the default HTML

# Example HTML (message_view.html - This would be a separate file)
# <h1>Message View</h1>
# <p>Send a message:</p>
# <form method="POST">
#   <input type="text" name="message">
#   <button type="submit">Send</button>
# </form>
#
# <p><b>Recent Messages:</b></p>
# <!--  Display messages here (e.g., using Django's template tags) -->
#

File: code-awesome/go-gin-react/super_dating/backend/src/pages/message.go
package pages

import (
	"fmt"
	"net/http"

	"github.com/gin-gonic/govigo"
)

type MessagePage struct {
	govigo.GovigoPage
}

func (m *MessagePage) HandleRequest(c *govigo.Context) {
	// Check if the request method is POST
	if c.Request.Method == "POST" {
		// Get the message content from the request body
		messageContent := c.Request.FormValue("message")

		// Validate the message content (basic check - length)
		if len(messageContent) > 0 {
			// Save the message to the database (simulated)
			m.saveMessageToDatabase(messageContent)

			// Redirect to the message list page
			c.Redirect(http.StatusSeeOther, "/messages")
		} else {
			// Handle invalid message content (e.g., display an error message)
			c.Error("Message cannot be empty")
		}
	} else if c.Request.Method == "GET" {
		// Handle GET requests (e.g., display a form to send a message)
		c.HTML("<h1>Send a Message</h1>", "Enter your message below:")
	} else {
		// Handle other request methods (e.g., return a 405 Method Not Allowed error)
		c.Error("Method Not Allowed")
	}
}

// Simulate saving the message to the database
func (m *MessagePage) saveMessageToDatabase(message string) {
	fmt.Printf("Saving message to database: %s\n", message)
	// In a real application, this function would connect to the database
	// and store the message there.

	// Simulate a potential vulnerability - Content Injection
	if message := "<h1>Content Injection Vulnerability</h1>"; containsContentInjection(message) {
		fmt.Println("WARNING: Potential content injection attempt detected!")
	}
}

// Helper function for demonstration purposes - simulating content injection
func containsContentInjection(text string) bool {
	// This is a very simplified example.  Real-world applications need more robust
	// input validation and sanitization.
	return true // Always return true for demonstration purposes.  In a real app, check for <script> or HTML tags.
}